File: plusone/src/services/postService.ts
import { api } from "./http";
import type { Post } from "../types/post";
import type { ProfileResponse } from "../types/profile";

export type AttendeeSummary = {
  id: string;
  firstName: string;
  lastName: string;
};

export const postService = {
  async getProfile(userId: string): Promise<ProfileResponse> {
    const { data } = await api.get<ProfileResponse>(`/users/${userId}/profile`);
    return data;
  },
  async list(userId: string): Promise<Post[]> {
    const { data } = await api.get<Post[]>("/posts", { params: { userId } });
    return data;
  },
  async create(post: Post): Promise<Post> {
    const { data } = await api.post<Post>("/posts", post);
    return data;
  },
  async update(id: string, post: Post): Promise<Post> {
    const { data } = await api.put<Post>(`/posts/${id}`, post);
    return data;
  },
  async remove(id: string): Promise<void> {
    await api.delete(`/posts/${id}`);
  },
  
  async getBookmarkedPosts(userId: string): Promise<Post[]> {
  const { data } = await api.get<Post[]>("/posts/bookmarked", {
    params: { userId },
  });
  return data;
},

  async bookmarkPost(userId: string, postId: string): Promise<void> {
    await api.post(`/posts/${postId}/bookmark`, null, {
      params: { userId },
    });
  },

  async unbookmarkPost(userId: string, postId: string): Promise<void> {
    await api.delete(`/posts/${postId}/bookmark`, {
      params: { userId },
    });
  },

  async rsvp(postId: string, userId: string): Promise<Post> {
    const { data } = await api.post<Post>(`/posts/${postId}/rsvp`, null, {
      params: { userId },
    });
    return data;
  },

  async cancelRsvp(postId: string, userId: string): Promise<Post> {
    const { data } = await api.delete<Post>(`/posts/${postId}/rsvp`, {
      params: { userId },
    });
    return data;
  },

  async getRsvps(postId: string, requestingUserId: string): Promise<AttendeeSummary[]> {
    const { data } = await api.get<AttendeeSummary[]>(`/posts/${postId}/rsvps`, {
      params: { requestingUserId },
    });
    return data;
  },
};

File: PlusOneBackend/src/main/java/com/plusone/PlusOneBackend/service/PostService.java
package com.plusone.PlusOneBackend.service;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.plusone.PlusOneBackend.model.Post;
import com.plusone.PlusOneBackend.model.User;
import com.plusone.PlusOneBackend.repository.PostRepository;
import com.plusone.PlusOneBackend.repository.UserRepository;

@Service
public class PostService {

    private static final String CATEGORY_EVENTS = "Events";

    private final UserRepository userRepository;
    private final PostRepository postRepository;

    @Autowired
    public PostService(UserRepository userRepository,
                       PostRepository postRepository) {
        this.userRepository = userRepository;
        this.postRepository = postRepository;
    }

    public void bookmarkPost(String userId, String postId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found: " + userId));

        // Optional: ensure post exists
        postRepository.findById(postId)
                .orElseThrow(() -> new RuntimeException("Post not found: " + postId));

        List<String> bookmarks = user.getBookmarkedPostIds();
        if (bookmarks == null) {
            bookmarks = new ArrayList<>();
            user.setBookmarkedPostIds(bookmarks);
        }

        if (!bookmarks.contains(postId)) {
            bookmarks.add(postId);
        }

        userRepository.save(user);
    }

    public void unbookmarkPost(String userId, String postId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found: " + userId));

        List<String> bookmarks = user.getBookmarkedPostIds();
        if (bookmarks != null && bookmarks.remove(postId)) {
            userRepository.save(user);
        }
    }

    public List<Post> getBookmarkedPosts(String userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found: " + userId));

        List<String> ids = user.getBookmarkedPostIds();
        if (ids == null || ids.isEmpty()) {
            return new ArrayList<>();
        }

        return postRepository.findAllById(ids);
    }

    /**
     * Add the user to the RSVP list for an event post.
     */
    public Post rsvpToEvent(String userId, String postId) {
        User user = requireUser(userId);
        Post post = requireEventPost(postId);

        List<String> rsvps = post.getRsvpUserIds();
        if (!rsvps.contains(user.getId())) {
            rsvps.add(user.getId());
            post.setRsvpUserIds(rsvps);
            postRepository.save(post);
        }

        return post;
    }

    /**
     * Remove the user from the RSVP list for an event post.
     */
    public Post cancelRsvp(String userId, String postId) {
        requireUser(userId);
        Post post = requireEventPost(postId);

        List<String> rsvps = post.getRsvpUserIds();
        if (rsvps.remove(userId)) {
            postRepository.save(post);
        }

        return post;
    }

    /**
     * Return RSVP'd users for an event post (in RSVP order).
     */
    public List<Post.AuthorSummary> getRsvps(String postId, String requestingUserId) {
        Post post = requireEventPost(postId);
        if (requestingUserId != null && !requestingUserId.equals(post.getUserId())) {
            throw new RuntimeException("Only the author can view RSVPs for this event");
        }

        List<String> attendeeIds = post.getRsvpUserIds();
        if (attendeeIds.isEmpty()) {
            return List.of();
        }

        Map<String, User> usersById = userRepository.findAllById(attendeeIds).stream()
                .collect(Collectors.toMap(User::getId, user -> user));

        return attendeeIds.stream()
                .map(usersById::get)
                .filter(Objects::nonNull)
                .map(user -> new Post.AuthorSummary(user.getId(), user.getFirstName(), user.getLastName()))
                .collect(Collectors.toList());
    }

    private Post requireEventPost(String postId) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new RuntimeException("Post not found: " + postId));

        if (!isEventsCategory(post)) {
            throw new RuntimeException("RSVPs are only supported for Events posts");
        }

        // Initialize RSVP list for older documents that might not have the field set yet.
        post.getRsvpUserIds();

        return post;
    }

    private boolean isEventsCategory(Post post) {
        return post.getCategory() != null && post.getCategory().equalsIgnoreCase(CATEGORY_EVENTS);
    }

    private User requireUser(String userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found: " + userId));
    }
}
